examples/systemcalls/systemcalls.c

```diff
@@ -92,7 +140,45 @@ bool do_exec_redirect(const char *outputfile, int count, ...)
  *   The rest of the behaviour is same as do_exec()
  *
 */
-
+    int res;    
+    pid_t pid;
+    mode_t mode = S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH ;
+    int fd =  open(outputfile, O_WRONLY|O_CREAT|O_TRUNC, mode );
+     if(fd == -1)
+     {
+	     perror("open");
+             abort();
+     }
+    pid=fork();
+    if(pid == -1){
+        perror("fork");
+        exit (-1);
+    }
+    else if(pid == 0){
+	    if (dup2(fd, 1) < 0)
+	    {
+		    perror("dup2");
+		    abort();
+	    }
+            int ret_val=execv(command[0],command);
+            if(ret_val == -1){
+                perror("execv");
+                exit(-1);
+            }
+    }
+    else{
+   	 pid_t id = waitpid(pid,&res,0);
+         if(id == -1){
+                perror("waitpid");
+                return false;
+	 }
+         if(WIFEXITED(res)){
+                if(WEXITSTATUS(res)!=0){
+                        return false;
```

Comment:
Missing cleanup in these return paths. 
Also see discussion in class about avoiding duplication of logic on these two very similar functions.